docs/src/concepts.md

```diff
@@ -0,0 +1,137 @@
+either directly or through a wrapper like `runForever()` or `waitFor()`.
+
+Each call to poll handles any file descriptors, timers and callbacks that are
+ready to be processed.
+
+Using `waitFor`, the result of a single asynchronous operation can be obtained:
+
+```nim
+proc myApp() {.async.} =
+  echo "Waiting for a second..."
+  await sleepAsync(1.seconds)
+  echo "done!"
+
+waitFor myApp()
+```
+
+It is also possible to keep running the event loop forever using `runForever`:
+
+```nim
+proc myApp() {.async.} =
+  while true:
+    await sleepAsync(1.seconds)
+    echo "A bit more than a second passed!"
+
+let future = myApp()
+runForever()
+```
+
+Such an application never terminates, thus it is rare that applications are
+structured this way.
+
+```admonish warning
+Both `waitFor` and `runForever` call `poll` which offers fine-grained
+control over the event loop steps.
+
+Nested calls to `poll`, `waitFor` and `runForever` are not allowed.
+```
+
+## Cancellation
+
+Any pending `Future` can be cancelled. This can be used for timeouts, to start
+multiple operations in parallel and cancel the rest as soon as one finishes,
+to initiate the orderely shutdown of an application etc.
+
+```nim
+{{#include ../examples/cancellation.nim}}
+```
+
+Even if cancellation is initiated, it is not guaranteed that the operation gets
+cancelled - the future might still be completed or fail depending on the
+order of events in the dispatcher and the specifics of the operation.
+
+If the future indeed gets cancelled, `await` will raise a
+`CancelledError` as is likely to happen in the following example:
+
+```nim
+proc c1 {.async.} =
+  echo "Before sleep"
+  try:
+    await sleepAsync(10.minutes)
+    echo "After sleep" # not reach due to cancellation
+  except CancelledError as exc:
+    echo "We got cancelled!"
+    # `CancelledError` is typically re-raised to notify the caller that the
+    # operation is being cancelled
+    raise exc
+
+proc c2 {.async.} =
+  await c1()
+  echo "Never reached, since the CancelledError got re-raised"
+
+let work = c2()
+waitFor(work.cancelAndWait())
+```
+
+The `CancelledError` will now travel up the stack like any other exception.
+It can be caught and handled (for instance, freeing some resources)
+
+Cancelling an already-finished `Future` has no effect, as the following example
+of downloading two web pages concurrently shows:
+
+```nim
+{{#include ../examples/twogets.nim}}
+```
+
+## Compile-time configuration
+
+`chronos` contains several compile-time [configuration options](./chronos/config.nim) enabling stricter compile-time checks and debugging helpers whose runtime cost may be significant.
+
+Strictness options generally will become default in future chronos releases and allow adapting existing code without changing the new version - see the [`config.nim`](./chronos/config.nim) module for more information.
+
+## Platform independence
+
+Several functions in `chronos` are backed by the operating system, such as
```

Comment:
I think this paragraph has outdated information, right now most of the functions which create sockets, files and network events returning either `Result[T, E]` or `bool` but did raise only `Defect`s.

Reply:
81a6584a44c69c03c47117c848dbbe63fac79fb5